test/data/data_tx_test.go

```diff
@@ -0,0 +1,527 @@
+	portMutex.Lock()
+	defer portMutex.Unlock()
+
+	node = baseNodePort
+	peer = basePeerPort
+	baseNodePort += num
+	basePeerPort += num
+
+	return
+}
+
+func TestDataTx(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(1)
+	setupConfig := &setup.Config{
+		NumberOfServers:     1,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+	leaderIndex := -1
+	require.Eventually(t, func() bool {
+		leaderIndex = c.AgreedLeader(t, 0)
+		return leaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	s := c.Servers[0]
+
+	// creating two databases -- db1 and db2
+	setup.CreateDatabases(t, s, []string{"db1", "db2"})
+
+	// creating two users -- alice and bob
+	aliceCert, _ := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "alice")
+	bobCert, _ := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "bob")
+	setup.CreateUsers(t, s, []*types.UserWrite{
+		{
+			User: &types.User{
+				Id:          "alice",
+				Certificate: aliceCert.Raw,
+				Privilege: &types.Privilege{
+					DbPermission: map[string]types.Privilege_Access{
+						"db1": types.Privilege_ReadWrite,
+						"db2": types.Privilege_ReadWrite,
+					},
+				},
+			},
+		},
+		{
+			User: &types.User{
+				Id:          "bob",
+				Certificate: bobCert.Raw,
+				Privilege: &types.Privilege{
+					DbPermission: map[string]types.Privilege_Access{
+						"db1": types.Privilege_ReadWrite,
+						"db2": types.Privilege_ReadWrite,
+					},
+				},
+			},
+		},
+	})
+
+	t.Run("blind write to a single database", func(t *testing.T) {
+		dataTx := &types.DataTx{
+			MustSignUserIds: []string{
+				"alice",
+			},
+			TxId: uuid.New().String(),
+			DbOperations: []*types.DBOperation{
+				{
+					DbName: "db1",
+					DataWrites: []*types.DataWrite{
+						{
+							Key:   "key1",
+							Value: []byte("key1value1"),
+						},
+						{
+							Key:   "key2",
+							Value: []byte("key2value1"),
+							Acl: &types.AccessControl{
+								ReadWriteUsers: map[string]bool{"alice": true},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		aliceSigner, err := s.Signer("alice")
+		require.NoError(t, err)
+
+		_, err = s.SubmitTransaction(t, constants.PostDataTx, &types.DataTxEnvelope{
+			Payload:    dataTx,
+			Signatures: map[string][]byte{"alice": testutils.SignatureFromTx(t, aliceSigner, dataTx)},
+		})
+		require.NoError(t, err)
+
+		for _, keyIndex := range []string{"1", "2"} {
+			resp, err := s.QueryData(t, "db1", "key"+keyIndex, "alice")
+			require.NoError(t, err)
+			require.Equal(t, []byte("key"+keyIndex+"value1"), resp.GetResponse().GetValue())
+		}
+	})
+
+	t.Run("read-write-delete to a single database", func(t *testing.T) {
```

Comment:
An additional test case is to check the "insert if does not exist" idiom, that is, when the read key version is nil (or empty? <0,0>)
It should insert only if the key does not exist and fail otherwise.
I think it only works if Version=nil but not if Version=<0,0>; please check that. 
I think that <0,0> should be semantically equivalent to nil in this case; what do you think?

Reply:
Done